File: clover-android-sdk-examples/src/main/java/com/clover/android/sdk/examples/MerchantDevicesV2TestActivity.java
package com.clover.android.sdk.examples;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.LoaderManager;
import android.content.CursorLoader;
import android.content.Loader;
import android.database.Cursor;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.ListView;
import android.widget.TextView;

import com.clover.sdk.v3.device.Device;
import com.clover.sdk.v3.merchant.MerchantDevicesV2Connector;
import com.clover.sdk.v3.merchant.MerchantDevicesV2Contract;

/**
 * Test activity that exercises the contract {@link MerchantDevicesV2Contract} and the connector
 * {@link MerchantDevicesV2Connector}.
 */
public class MerchantDevicesV2TestActivity extends Activity implements LoaderManager.LoaderCallbacks<Cursor> {

  private ListView listView;
  private MerchantDevicesV2Adapter adapter;
  private View progressView;
  private View contentView;
  private TextView thisDeviceView;
  private TextView thisSerialView;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(R.layout.activity_merchant_devices_v2);

    progressView = findViewById(R.id.progress_bar);
    contentView = findViewById(R.id.content_layout);

    thisSerialView = findViewById(R.id.this_serial);
    thisDeviceView = findViewById(R.id.this_device);

    listView = findViewById(R.id.list_view);
    listView.setEmptyView(findViewById(R.id.list_view_empty));
    adapter = new MerchantDevicesV2Adapter(this);
    listView.setAdapter(adapter);
  }

  @SuppressLint("StaticFieldLeak")
  @Override
  protected void onStart() {
    super.onStart();

    progressView.setVisibility(View.VISIBLE);
    contentView.setVisibility(View.GONE);

    getLoaderManager().initLoader(0, null, this);

    new AsyncTask<Void,Void,String>() {
      @Override
      protected String doInBackground(Void... voids) {
        return new MerchantDevicesV2Connector(getApplicationContext()).getSerial();
      }

      @Override
      protected void onPostExecute(String serial) {
        if (serial == null) {
          return;
        }
        thisSerialView.setText(serial);
      }
    }.execute();
    new AsyncTask<Void,Void,Device>() {
      @Override
      protected Device doInBackground(Void... voids) {
        return new MerchantDevicesV2Connector(getApplicationContext()).getDevice();
      }

      @Override
      protected void onPostExecute(Device device) {
        if (device == null) {
          return;
        }
        thisDeviceView.setText(String.format("%s: %s %s",
            device.getProductName(), device.getModel(), device.getSerial()));
      }
    }.execute();

  }

  @Override
  public Loader<Cursor> onCreateLoader(int id, Bundle args) {
    return new CursorLoader(this, MerchantDevicesV2Contract.Device.CONTENT_URI, null, null, null, null);
  }

  @Override
  public void onLoadFinished(Loader loader, Cursor data) {
    adapter.swapCursor(data);

    progressView.setVisibility(View.GONE);
    contentView.setVisibility(View.VISIBLE);
  }

  @Override
  public void onLoaderReset(Loader loader) {
    // we're not binding to an adapter or hanging on to the cursor, so no need to do anything here
  }
}

File: clover-android-sdk-examples/src/main/java/com/clover/android/sdk/examples/MerchantDevicesV2TestActivity.kt
package com.clover.android.sdk.examples

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.loader.content.CursorLoader
import com.clover.android.sdk.examples.databinding.ActivityMerchantDevicesV2Binding
import com.clover.sdk.v3.merchant.MerchantDevicesV2Connector
import com.clover.sdk.v3.merchant.MerchantDevicesV2Contract
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Test activity that exercises the contract [MerchantDevicesV2Contract] and the connector
 * [MerchantDevicesV2Connector].
 */
class MerchantDevicesV2TestActivity : AppCompatActivity() {
  private lateinit var adapter: MerchantDevicesV2Adapter
  private lateinit var binding: ActivityMerchantDevicesV2Binding
  private lateinit var connector: MerchantDevicesV2Connector

  protected override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    connector = MerchantDevicesV2Connector(applicationContext)

    binding = ActivityMerchantDevicesV2Binding.inflate(layoutInflater)
    setContentView(binding.root)

    binding.listView.setEmptyView(findViewById<View?>(R.id.list_view_empty))
    adapter = MerchantDevicesV2Adapter(this)
    binding.listView.setAdapter(adapter)

    binding.reboot.setOnClickListener {
      val now = binding.rebootNow.isChecked
      lifecycleScope.launch {
        withContext(Dispatchers.IO) { connector.reboot("User requested reboot", now) }
      }
    }
  }

  @SuppressLint("StaticFieldLeak")
  override fun onStart() {
    super.onStart()

    binding.progressBar.visibility = View.VISIBLE
    binding.contentLayout.visibility = View.GONE

    lifecycleScope.launch {
      val device = withContext(Dispatchers.IO) { connector.device }
      binding.thisSerial.text = device.serial
      binding.thisModel.text = device.model
      binding.thisName.text = device.productName

      val simIccId = withContext(Dispatchers.IO) { connector.simIccId }
      binding.thisSimmIccId.text = simIccId ?: "N/A"

      CursorLoader(
          this@MerchantDevicesV2TestActivity,
          MerchantDevicesV2Contract.Device.CONTENT_URI,
          null,
          null,
          null,
          null
      ).apply {
        adapter.swapCursor(withContext(Dispatchers.IO) { loadInBackground() })
        binding.progressBar.visibility = View.GONE
        binding.contentLayout.visibility = View.VISIBLE
      }
    }
  }
}
